resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala

```diff
@@ -63,4 +63,66 @@ object KubernetesFeaturesTestUtils {
   def containerHasEnvVar(container: Container, envVarName: String): Boolean = {
     container.getEnv.asScala.exists(envVar => envVar.getName == envVarName)
   }
+
+  def containerHasEnvVars(container: Container, envs: Map[String, String]): Unit = {
+    assertHelper[Set[(String, String)]](envs.toSet,
+      container.getEnv.asScala
+        .map { e => (e.getName, e.getValue) }.toSet,
+      subsetOfTup[Set[(String, String)], String], "a subset of")
+  }
+
+  def containerHasVolumeMounts(container: Container, vms: Map[String, String]): Unit = {
+    assertHelper[Set[(String, String)]](vms.toSet,
+      container.getVolumeMounts.asScala
+        .map { vm => (vm.getName, vm.getMountPath) }.toSet,
+      subsetOfTup[Set[(String, String)], String], "a subset of")
+  }
+
+  def podHasLabels(pod: Pod, labels: Map[String, String]): Unit = {
+    assertHelper[Set[(String, String)]](labels.toSet, pod.getMetadata.getLabels.asScala.toSet,
+      subsetOfTup[Set[(String, String)], String], "a subset of")
+  }
+
+  def podHasVolumes(pod: Pod, volumes: Seq[Volume]): Unit = {
+    assertHelper[Set[Volume]](volumes.toSet, pod.getSpec.getVolumes.asScala.toSet,
+      subsetOfElem[Set[Volume], Volume], "a subset of")
+  }
+
+  // Mocking bootstrapHadoopConfDir
+  def hadoopConfBootPod(inputPod: SparkPod): SparkPod =
+    SparkPod(
+      new PodBuilder(inputPod.pod)
+        .editOrNewMetadata()
+          .addToLabels("bootstrap-hconf", "true")
+          .endMetadata()
+        .build(),
+      inputPod.container)
+
+  // Mocking bootstrapKerberosPod
+  def krbBootPod(inputPod: SparkPod): SparkPod =
+    SparkPod(
+      new PodBuilder(inputPod.pod)
+        .editOrNewMetadata()
+          .addToLabels("bootstrap-kerberos", "true")
+          .endMetadata()
+        .build(),
+      inputPod.container)
+
+  // Mocking bootstrapSparkUserPod
+  def userBootPod(inputPod: SparkPod): SparkPod =
+    SparkPod(
+      new PodBuilder(inputPod.pod)
+        .editOrNewMetadata()
+          .addToLabels("bootstrap-user", "true")
+          .endMetadata()
+        .build(),
+      inputPod.container)
+
+  def subsetOfElem[T <: Set[B], B <: Any]: (T, T) => Boolean = (a, b) => a.subsetOf(b)
+  def subsetOfTup[T <: Set[(B, B)], B <: Any]: (T, T) => Boolean = (a, b) => a.subsetOf(b)
+
+  def assertHelper[T](con1: T, con2: T,
+      expr: (T, T) => Boolean = (a: T, b: T) => a == b, exprMsg: String = "equal to"): Unit = {
+    assert(expr(con1, con2), s"$con1 is not $exprMsg $con2 as expected")
```

Comment:
This interaction seems really awkward to me. Why can't we just embed the error message directly into the assertion on each assertion statement? I'm also not familiar enough with the ScalaTest framework but perhaps we can explore other options the library has to offer. (In Java I'm used to using [AssertJ](https://github.com/joel-costigliola/assertj-core) for fluent assertions but I'm not sure if that will be of any help for Scala objects, particularly Scala collections without converting everything.)

Reply:
I thought it would be better than doing a custom string at every assert statement. I don't find it to be too awkward, but *shrug* :) I actually kind of like it since we could do a check with any expression

Reply:
I think here the simpler approach is preferred - on each assertion line I want to see right then and there what error message I'll get back, not having to go through a helper method.